mantidproject/mantid
# Fix performance regression in V3D

Commit 48d17da in PR #24677 refactored the `Kernel::V3D` class enabling `constexpr` `V3D`s. However, while overall being faster than before, the commit caused a regression in `testElementAccessOperator` performance test.

This PR fixes the performance regression by changing the private representation of `V3D`'s elements from three `double` values to `std::array<double, 3>`, in line with what we have in `V2D`. The speed of the other `V3D` performance tests should be the same as before.

**`V3D` performance test duration** on author's system (Ubuntu 16.04):

(`m_sampleSize = 1000000000` on all tests below)

| Branch                    | `testElementAccessOperator` | All performance tests |
| ------------------------- | --------------------------- | --------------------- |
| This                      | 1.0s                        | 6.6s                  |
| Current master            | 2.7s                        | 8.2s                  |
| Master before 48d17da (*) | 1.0s                        |  20s                  |

(*) commit ef15643

**Report to:** @DanNixon 

**To test:**

Build in release mode and time the `V3DTestPerformance` unit tests. Compare with current master --- note that you should add one more zero to the `m_sampleSize` on master, see line 587 in `V3DTest.h`. This PR should not perform any worse, at least.

Fixes #24860.

*This does not require release notes* because it is an internal change only.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
